### DISCUSSION TITLE
How to launch shell amd64 on ARM

### DISCUSSION THREAD

**Discussion by OlegGulevskyy** [lima-vm/lima#1043] [Q&A]
Very sorry for a what probably is very obvious question for some of you, but I am trying to run some servers on my mac M1, and they can only be run on Intel
I am trying to follow the readme here:
https://github.com/lima-vm/lima/blob/master/docs/multi-arch.md#fast-mode
Slow mode is very very slow,I could not wait until the end and stopped the process.
Fast mode is - I am getting x86_64 in the console as result when I use command
`lima nerdctl run --platform=amd64 --rm alpine uname -m`
However, I don't know where to go from here. How can I open op a shell and start working with my directory and files as if I am in x86_64 architecture?
The readme file stops on those commands, so I am assuming either it's very obvious or not possible.

Thank you!

**Comment by OlegGulevskyy** [comment]
I tried running the following
`lima nerdctl run -it --platform=amd64 --rm alpine bash`
But got into error
```
FATA[0000] failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "bash": executable file not found in $PATH: unknown
```

**Reply by AkihiroSuda** [reply]
sh, not bash

**Reply by OlegGulevskyy** [reply]
oh yes, it does work with `sh` thank you!
when I do so, I can see that there are no files shared with this instance. Is this because it's an image that has no files added to it? Because when I use simply `lima` I am sort of logged to something where I can access all my files. Is this possible to achieve the same?
Not sure how much sense I make when I am trying to explain this, sorry if this is not clear, this topic is very new to me and I am trying to wrap my head around it.

**Reply by afbjorklund** [reply]
If you are running applications that are **not** containerized, you can install the same level of support by issuing command:

```
sudo apt install qemu binfmt-support qemu-user-static
```

This will allow the (x86_64) VM to run the (aarch64) ELF, by calling QEMU to emulate the application (but **not** the OS)

Before:

```console
anders@lima-default:/home/anders$ ./hello-amd64 
Hello, World!
anders@lima-default:/home/anders$ ./hello-arm64 
bash: ./hello-arm64: cannot execute binary file: Exec format error
```

After:
```console
anders@lima-default:/home/anders$ ./hello-amd64 
Hello, World!
anders@lima-default:/home/anders$ ./hello-arm64 
Hello, World!
```

When trying to run two binaries, for different CPU architectures:

```
hello-amd64: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, BuildID[sha1]=00cc4b6f2c0973f0d0cfedc53981db8286759fdd, for GNU/Linux 3.2.0, not stripped
hello-arm64: ELF 64-bit LSB executable, ARM aarch64, version 1 (GNU/Linux), statically linked, BuildID[sha1]=0482bf10f3f4ba1d6ae9b86e80179d5503b185ba, for GNU/Linux 3.7.0, not stripped
```

See https://wiki.debian.org/QemuUserEmulation for more details about this Linux kernel mechanism ([binfmt_misc](https://en.wikipedia.org/wiki/Binfmt_misc)).

Note that if you have dynamically linked applications (not static), then you will have the usual dependency issues:

```
/lib64/ld-linux-x86-64.so.2: No such file or directory
/lib/ld-linux-aarch64.so.1: No such file or directory
```

This is why people want to run them in containers, since it provides the dependencies - like a modern `chroot`...

PS. In your case, it would be the opposite (trying to run a x86_64 binary on aarch64, and x86_64-linux-gnu/libc.so.6)
